src/Compilers/Core/Portable/ConstantValueSpecialized.cs

```diff
@@ -753,6 +771,11 @@ private sealed class ConstantValueSingle : ConstantValueDiscriminated
             public ConstantValueSingle(double value)
                 : base(ConstantValueTypeDiscriminator.Single)
             {
+                if (double.IsNaN(value))
+                {
+                    value = _s_IEEE_canonical_NaN;
```

Comment:
This is relying on the underlying hardware to exactly preserve both the sign and the "payload" bits when converting from double to single, which is not required by the spec.

Reply:
It is required by the spec that a float canonical NaN is required to round trip (without loss) to and from a double NaN [IEEE Std 754-2008 section 6.2.3].

Reply:
Right but 6.3 explicitly specifies that for both inputs and results, the standard does not interpret the sign of the NaN. And 10.4 specifies that changing the sign of a NaN is allowed as a "value-changing transformation which preserves the literal meaning of the source code".